third_party/alerter/alerter/AppDelegate.m

```diff
@@ -254,345 +193,237 @@ - (void)deliverNotificationWithTitle:(NSString *)title
                              options:(NSDictionary *)options
                                sound:(NSString *)sound;
 {
-    // First remove earlier notification with the same group ID.
-    if (options[@"groupID"]) [self removeNotificationWithGroupID:options[@"groupID"]];
-    
-    NSUserNotification *userNotification = [NSUserNotification new];
-    userNotification.title = title;
-    userNotification.subtitle = subtitle;
-    userNotification.informativeText = message;
-    userNotification.userInfo = options;
-    
-    
-    
-    if(isMavericks()){
-        // Mavericks options
-        if(options[@"appIcon"]){
-            // replacement app icon
-            [userNotification setValue:[self getImageFromURL:options[@"appIcon"]] forKey:@"_identityImage"];
-            [userNotification setValue:@(false) forKey:@"_identityImageHasBorder"];
-        }
-        if(options[@"contentImage"]){
-            // content image
-            userNotification.contentImage = [self getImageFromURL:options[@"contentImage"]];
-        }
-    }
-    
-    // Actions
-    if (options[@"actions"]){
-        [userNotification setValue:@YES forKey:@"_showsButtons"];
-        NSArray *myActions = [options[@"actions"] componentsSeparatedByString:@","];
-        if (myActions.count > 1) {
-            [userNotification setValue:@YES forKey:@"_alwaysShowAlternateActionMenu"];
-            [userNotification setValue:myActions forKey:@"_alternateActionButtonTitles"];
-            
-            //Main Actions Title
-            if(options[@"dropdownLabel"]){
-                userNotification.actionButtonTitle = options[@"dropdownLabel"];
-                userNotification.hasActionButton = 1 ;
-            }
-        }else{
-            userNotification.actionButtonTitle = options[@"actions"];
-        }
-    }else if (options[@"reply"]) {
-        [userNotification setValue:@YES forKey:@"_showsButtons"];
-        userNotification.hasReplyButton = 1;
-        userNotification.responsePlaceholder = options[@"reply"];
-    }
-    
-    // Close button
-    if(options[@"closeLabel"]){
-        userNotification.otherButtonTitle = options[@"closeLabel"];
+  if (options[@"groupID"] && [self notificationWithGroupIdExists:options[@"groupID"]]) {
+    NSLog(@"Sending notification with the same groupID as existing ones is not supported.");
```

Comment:
Can you say more about this switch from removing old notifications with a matching GroupID to exiting ? Is it just not implemented or is there another issue as well that makes it unsupported?

Reply:
We still support removing old notifications with a matching GroupID, implemented at line 154-157. What we don't support here was sending a notification with the same groupID as an existing one. 

In the old alerter, when sending a notification with the same groupID as an existing one, the existing one will be removed before the new one get displayed. However, for us it will make monitoring notification removal subject to race conditions (concurrent editing of an object tracking notifications and removals). A careful design with some locks might eliminate these race conditions, but currently we don't actually need this feature.

Note that if we don't use a tracking object to track notifications and removals, but instead periodically check whether the corresponding notification is still present, as what is done in the old alerter, then the above race conditions could be eliminated. However, that will potentially cause new race conditions since it is unknown whether the notification action callback will be called before the notification disappearance being detected.

Reply:
I just realized I could maybe combine the tracking object with periodically checking whether the corresponding notification is still present. So when the notification disappears, the detector will know whether it was caused by removal.

Reply:
The case I was thinking of was two triggers within a Data Collector that overlapped, for example a Schedule and an Event Trigger. We would want to ensure that the latest one wins. I think. So, would that be the case for removal of an existing for the same Group ID?

Reply:
During my trying (to enable sending notifications with the same groupID), I sometimes encounter an issue: when I consecutively and quickly send multiple notifications with the same groupID and quickly click the notification of the last one, the alerter will hang... So for now let's just keep it as is.

Reply:
Is this ready for checkin? Or are you doing other work on it still?

Reply:
Sorry for the delayed updates. It is not ready yet. I was still working on the bug we discovered this morning.